src/esw/can_node/CanNode.hpp

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <cstdio>
+#include <cstring>
+#include <iostream>
+#include <linux/can.h>
+#include <linux/can/raw.h>
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <vector>
+
+class CanNode {
+public:
+    CanNode() {
+        struct sockaddr_can addr {};
+        struct ifreq ifr {};
+
+        if ((s = socket(PF_CAN, SOCK_RAW, CAN_RAW)) < 0) {
+            perror("Error while opening socket");
+        }
+
+        const char* ifname = "can0";
+        strcpy(ifr.ifr_name, ifname);
+        ioctl(s, SIOCGIFINDEX, &ifr);
+
+        addr.can_family = AF_CAN;
+        addr.can_ifindex = ifr.ifr_ifindex;
+
+        if (bind(s, (struct sockaddr*) &addr, sizeof(addr)) < 0) {
+            perror("Error in socket bind");
+        }
+    }
+
+    void send_frame() const {
+        // write(s, &frame, sizeof(struct can_frame));
+        sendto(s, &frame, sizeof(struct can_frame), 0, nullptr, 0);
+    }
+
+    void set_bus(uint8_t bus) {
+        this->bus = bus;
+    }
+
+    void set_frame_id(uint16_t id) {
+        uint32_t identifier = static_cast<uint32_t>(id) & 0x000007FF; // can_id[0-10]
```

Comment:
How did you get these magic numbers?

Reply:
So the socketcan can_id is a uint32_t with format:
[0-28]: CAN identifier (11/29bit)
[29]: Error frame flag (0 = data frame, 1 = error frame)
[30]: Remote transmission request flag (1 = rtr frame)
[31]: Frame format flag (0 = standard 11bit, 1 = extended 29bit)

I'm just defining each section of the id as separate bit masks in order to lay out all of the separate bits we have to set. I've added additional comments explaining this. I've also added logic to implement both extended and standard frame identifier bits. Will push soon.

Reply:
Ok, makes sense - see my comments aobut using std::bitset, may make your life easier: https://en.cppreference.com/w/cpp/utility/bitset